homeassistant/components/nice_go/coordinator.py

```diff
@@ -178,16 +183,26 @@ async def _update_refresh_token(self) -> None:
 
     async def client_listen(self) -> None:
         """Listen to the websocket for updates."""
-        self.api.event(self.on_connected)
-        self.api.event(self.on_data)
-        try:
-            await self.api.connect(reconnect=True)
-        except ApiError:
-            _LOGGER.exception("API error")
+        self._unsub_connected = self.api.listen("on_connected", self.on_connected)
+        self._unsub_data = self.api.listen("on_data", self.on_data)
+        self._unsub_connection_lost = self.api.listen(
+            "on_connection_lost", self.on_connection_lost
+        )
+
+        for attempt in range(RECONNECT_ATTEMPTS):
+            if self.hass.is_stopping:
+                return
+
+            try:
+                await self.api.connect(reconnect=True)
+            except ApiError as e:
+                _LOGGER.exception("API error")
+                if attempt == RECONNECT_ATTEMPTS - 1:
+                    raise ConfigEntryNotReady from e
```

Comment:
Move raising after the loop, in that case no check is needed.
Buid in a 5 sec delay when there is an exception. The delay time should be a const, so you can mock in in the tests.

Reply:
`client_listen` is a background task, so `ConfigEntryNotReady` won't propagate iirc. Should I call `self.async_set_update_error` or raise anyway? Raising doesn't even show in logs

Reply:
Ah, okay. May be just log then, and exit after it went wrong for a longer time?

If you want to keep the loop going, unless hass is stopping, you should way a longer time before retrying. Not sure the API has rate limits.